include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
+        } else {
+            async_read_some_op<MutableBufferSequence, Handler> operation{
+                *this, sequence, std::move(init.completion_handler)};
+            next_layer().get_io_service().dispatch([operation = std::move(
+                                                        operation)]() mutable {
+                beast::error_code error{};
+                operation(error, 0);
+            });
+        }
+        return init.result.get();
+    }
+
+    template <typename ConstBufferSequence, typename Handler>
+    auto async_write_some(ConstBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        async_write_some_op<ConstBufferSequence, Handler> operation{
+            *this, sequence, std::move(init.completion_handler)};
+        next_layer().get_io_service().dispatch([operation = std::move(
+                                                    operation)]() mutable {
+            beast::error_code error{};
+            operation(error, 0);
+        });
+        return init.result.get();
+    }
+
+    void set_secrets(const crypto::aead_secrets& secrets) {
+        this->secrets = secrets;
+    }
+
+private:
+    template <typename ConstBufferSequence, typename Handler>
+    struct async_write_some_op : public asio::coroutine {
+        async_write_some_op(secure_socket& socket,
+                            const ConstBufferSequence& sequence,
+                            Handler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<Handler>(handler)} {}
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                static_cast<void>(n);
+                socket.encrypt_pdu(sequence, pdu);
+                CORO_YIELD asio::async_write(socket.next_layer(),
+                                             asio::buffer(pdu), *this);
+                handler(error, asio::buffer_size(sequence));
+            }
+        }
+
+        secure_socket& socket;
+        ConstBufferSequence sequence;
+        Handler handler;
```

Comment:
You have a huge problem here, because you have type-erased the handler without providing the necessary hooks (allocation, continuation, and invocation). You need to overload the four free function hooks for your coroutine, otherwise you will get undefined behaviors. Copy the ones from Beast, and modify them for your type:
https://github.com/boostorg/beast/blob/20a8f7d75b45f0ababc2547f50833b7f4ba636d3/include/boost/beast/core/detail/bind_handler.hpp#L137

Any time you write a composed operation that stores the caller's handler, you have to provide the hooks. This is going to change a little bit in the next Asio update, instead of those 4 functions you will just need to provide two functions, `get_executor` and `get_associated_allocator`. It will be easy.